modules/xfem/src/userobjects/MeshCut3DUserObject.C

```diff
@@ -0,0 +1,1160 @@
+  {
+    Point p(cut_point[0], cut_point[1], cut_point[2]);
+    Real dotp = ((p - p1) * (p2 - p1)) / ((p2 - p1) * (p2 - p1));
+    if (isInsideCutPlane(vertices, p) && dotp > 1)
+    {
+      pint = p;
+      has_intersection = true;
+    }
+  }
+  return has_intersection;
+}
+
+bool
+MeshCut3DUserObject::isInsideEdge(const Point & p1, const Point & p2, const Point & p) const
+{
+  Real dotp1 = (p1 - p) * (p2 - p1);
+  Real dotp2 = (p2 - p) * (p2 - p1);
+  return (dotp1 * dotp2 <= 0.0);
+}
+
+Real
+MeshCut3DUserObject::getRelativePosition(const Point & p1, const Point & p2, const Point & p) const
+{
+  // get the relative position of p from p1
+  Real full_len = (p2 - p1).norm();
+  Real len_p1_p = (p - p1).norm();
+  return len_p1_p / full_len;
+}
+
+bool
+MeshCut3DUserObject::isInsideCutPlane(const std::vector<Point> & vertices, Point p) const
+{
+  unsigned int n_node = vertices.size();
+
+  Point normal;
+  Point center;
+
+  findElemNormal(vertices, normal, center);
+
+  bool inside = false;
+  int counter = 0;
+
+  for (unsigned int i = 0; i < n_node; ++i)
+  {
+    unsigned int iplus1 = (i < n_node - 1 ? i + 1 : 0);
+    Point middle2p = p - 0.5 * (vertices[i] + vertices[iplus1]);
+    const Point side_tang = vertices[iplus1] - vertices[i];
+    Point side_norm = side_tang.cross(normal);
+    Xfem::normalizePoint(middle2p);
+    Xfem::normalizePoint(side_norm);
+    if (middle2p * side_norm <= 0.0)
+      counter += 1;
+  }
+  if (counter == n_node)
+    inside = true;
+  return inside;
+}
+
+bool
+MeshCut3DUserObject::findElemNormal(const std::vector<Point> & vertices,
+                                    Point & normal,
+                                    Point & center) const
+{
+  unsigned int n_node = vertices.size();
+
+  center(0) = center(1) = center(2) = 0;
+  normal(0) = normal(1) = normal(2) = 0;
+
+  for (unsigned int i = 0; i < n_node; ++i)
+    center += vertices[i];
+  center /= n_node;
+
+  for (unsigned int i = 0; i < n_node; ++i)
+  {
+    unsigned int iplus1(i < n_node - 1 ? i + 1 : 0);
+    Point ray1 = vertices[i] - center;
+    Point ray2 = vertices[iplus1] - center;
+    normal += ray1.cross(ray2);
+  }
+  normal /= n_node;
+  Xfem::normalizePoint(normal);
+  return true; // to be done
+}
+
+void
+MeshCut3DUserObject::findBoundaryNodes()
+// this is a simple algorithm simply based on the added angle = 360 degrees
+// works fine for planar cutting surface
+// for curved cutting surface, need to re-work this subroutine to make it more general
+{
+  _boundary_map.clear();
+
+  unsigned int n_nodes = _cut_mesh->n_nodes();
+  std::vector<Real> angle(n_nodes, 0);
+  std::vector<unsigned int> node_id(_cut_elem_nnode);
+
+  for (unsigned int i = 0; i < n_nodes; ++i)
+    angle[i] = 0;
+
+  const MeshBase::element_iterator elem_begin = _cut_mesh->elements_begin();
+  MeshBase::element_iterator elem_end = _cut_mesh->elements_end();
+  MeshBase::element_iterator elem_it = elem_begin;
```

Comment:
You use `auto` a lot of other places. This is actually an ideal place to use it (to avoid typing these long iterator names).